codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java

```diff
@@ -474,46 +553,93 @@ public String booleanShape(final BooleanShape shape) {
+      writer.addImportFromModule(
+        SmithyNameResolver.getGoModuleNameForSdkNamespace(
+          shape.getId().getNamespace()
+        ),
+        "types",
+        SmithyNameResolver.smithyTypesNamespace(shape)
+      );
+    }
     if (shape.hasTrait(EnumTrait.class)) {
-      return """
-         return func () *%s.%s {
-         var u %s.%s
-                     if %s == nil {
-                         return nil
-                     }
-      	inputEnum := %s.(%s)
-      	index := -1;
-      	for allEnums := dafny.Iterate(%s{}.AllSingletonConstructors()); ; {
-      		enum, ok := allEnums()
-      		if ok {
-      			index++
-      			if enum.(%s).Equals(inputEnum) {
-      				break;
-      			}
-      		}
-      	}
-
-      	return &u.Values()[index]
-      }()""".formatted(
-          SmithyNameResolver.smithyTypesNamespace(shape),
-          context.symbolProvider().toSymbol(shape).getName(),
-          SmithyNameResolver.smithyTypesNamespace(shape),
-          context.symbolProvider().toSymbol(shape).getName(),
-          dataSource,
-          dataSource,
-          DafnyNameResolver.getDafnyType(
-            shape,
-            context.symbolProvider().toSymbol(shape)
-          ),
-          DafnyNameResolver.getDafnyCompanionStructType(
-            shape,
-            context.symbolProvider().toSymbol(shape)
-          ),
-          DafnyNameResolver.getDafnyType(
-            shape,
-            context.symbolProvider().toSymbol(shape)
-          )
-        );
+      if (this.isOptional) {
+        return """
+          return func () *%s.%s {
+          var u %s.%s
+          if %s == nil {
+              return nil
+          }
+        	inputEnum := %s.(%s)
+        	index := -1;
+        	for allEnums := dafny.Iterate(%s{}.AllSingletonConstructors()); ; {
+        		enum, ok := allEnums()
+        		if ok {
+        			index++
+        			if enum.(%s).Equals(inputEnum) {
+        				break;
+        			}
+        		}
+        	}
+
+        	return &u.Values()[index]
+        }()""".formatted(
+            SmithyNameResolver.smithyTypesNamespace(shape),
+            context.symbolProvider().toSymbol(shape).getName(),
+            SmithyNameResolver.smithyTypesNamespace(shape),
+            context.symbolProvider().toSymbol(shape).getName(),
+            dataSource,
+            dataSource,
+            DafnyNameResolver.getDafnyType(
+              shape,
+              context.symbolProvider().toSymbol(shape)
+            ),
+            DafnyNameResolver.getDafnyCompanionStructType(
+              shape,
+              context.symbolProvider().toSymbol(shape)
+            ),
+            DafnyNameResolver.getDafnyType(
+              shape,
+              context.symbolProvider().toSymbol(shape)
+            )
+          );
+      } else {
+        return """
+          return func () %s.%s {
+          var u %s.%s
+        	inputEnum := %s.(%s)
+        	index := -1;
```

Comment:
why does index start at -1? no 0?

Reply:
We assign `-1` so that while iterating if none of the values match (line 616) we don't end up returning first value in the enum (line 622), rather panic / throw exception.

Reply:
yup! as soon as i got up from my desk and walked around I realized why, nice